### PR TITLE
Support Array as argument and return value in R server

### PR DESCRIPTION
**Pull request by HaozhouWang**
1. Support protobuf array format as input argument for R function
2. Support convert R vector into protobuf array format
3. Current only 1-D array is supported
4. Code refactor
5. Unit tests updated

### CODE REVIEW

File: src/rtypeconverter.cc [hunk 1]
```diff
@@ -300,3 +299,222 @@ void ConvertToProtoBuf::compositeToProtoBuf(SEXP v, std::vector<PlcDataType> &su
...
+                if (!v.values()[i].isnull()) {
+                    SET_STRING_ELT(result, i,
+                                   COPY_TO_USER_STRING(v.values()[i].stringvalue().c_str()));
+                } else {
+                    SET_STRING_ELT(result, i, NA_STRING);
+                }
+            }
+        } break;
+        default:
+            this->rLog->log(RServerLogLevel::WARNINGS,
```
Comment (wengyanqing) [inline src/rtypeconverter.cc]: Why not error out here instead of just print WARNING ?

File: src/rtypeconverter.cc [hunk 2]
```diff
@@ -300,3 +299,222 @@ void ConvertToProtoBuf::compositeToProtoBuf(SEXP v, std::vector<PlcDataType> &su
...
+                }
+            }
+        } break;
+        default:
+            this->rLog->log(RServerLogLevel::WARNINGS, "Unsupport type in value %d", type);
+            break;
+    }
+}
+
+SEXP ConvertToSEXP::allocRVector(const plcontainer::PlcDataType &type, int length) {
```
Comment (wengyanqing) [inline src/rtypeconverter.cc]: PlcDataType is enum, so it's better to pass by value than const reference. There are several similar issues in other functions.

File: src/rtypeconverter.cc (hunk 1 above)
Comment (HaozhouWang) [reply src/rtypeconverter.cc]: ERROR is the unrecoverable issue and will lead rserver to exit. WARNING is the recoverable issue and will not lead rserver to exit.

File: src/rtypeconverter.cc (hunk 2 above)
Comment (HaozhouWang) [reply src/rtypeconverter.cc]: Thx, fixed.

**Pull request merged**